File: basic_dsa_thirty_three.py
"""
Basic DSA Problems - One More Fundamental Question
Simple and essential DSA problems for beginners
"""


# Problem: Find nth Prime Number
# Find the nth prime number in the sequence of prime numbers
# Example: 1st prime = 2, 2nd prime = 3, 3rd prime = 5, 4th prime = 7, 5th prime = 11

def is_prime(n):
    """
    Check if a number is prime
    Time Complexity: O(√n)
    Space Complexity: O(1)
    """
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    
    i = 3
    while i * i <= n:
        if n % i == 0:
            return False
        i += 2
    
    return True


def nth_prime_brute_force(n):
    """
    Find nth prime number using brute force
    Time Complexity: O(n * √n)
    Space Complexity: O(1)
    """
    if n <= 0:
        raise ValueError("n must be positive")
    
    if n == 1:
        return 2
    
    count = 1
    num = 3
    
    while count < n:
        if is_prime(num):
            count += 1
            if count == n:
                return num
        num += 2  # Check only odd numbers (except 2)
    
    return num - 2


def nth_prime_sieve_optimized(n):
    """
    Find nth prime number using optimized approach with Sieve of Eratosthenes
    Time Complexity: O(n log log n) for sieve
    Space Complexity: O(n)
    """
    if n <= 0:
        raise ValueError("n must be positive")
    
    if n == 1:
        return 2
    
    # Estimate upper bound: n * log(n) + n * log(log(n)) for n > 6
    if n <= 6:
        primes = [2, 3, 5, 7, 11, 13]
        return primes[n - 1]
    
    import math
    upper_bound = int(n * math.log(n) + n * math.log(math.log(n)))
    if upper_bound < 20:
        upper_bound = 20
    
    # Sieve of Eratosthenes
    is_prime_list = [True] * (upper_bound + 1)
    is_prime_list[0] = is_prime_list[1] = False
    
    for i in range(2, int(math.sqrt(upper_bound)) + 1):
        if is_prime_list[i]:
            for j in range(i * i, upper_bound + 1, i):
                is_prime_list[j] = False
    
    # Collect primes
    primes = []
    for i in range(2, upper_bound + 1):
        if is_prime_list[i]:
            primes.append(i)
            if len(primes) == n:
                return i
    
    # If not enough primes found, extend search
    return nth_prime_brute_force(n)


def get_first_n_primes(n):
    """
    Get list of first n prime numbers
    Time Complexity: O(n * √n)
    Space Complexity: O(n)
    """
    if n <= 0:
        return []
    
    primes = []
    num = 2
    
    while len(primes) < n:
        if is_prime(num):
            primes.append(num)
        num += 1
    
    return primes


def get_primes_in_range(start, end):
    """
    Get all prime numbers in a given range
    Time Complexity: O((end - start) * √n)
    Space Complexity: O(k) where k is the number of primes
    """
    primes = []
    
    for num in range(start, end + 1):
        if is_prime(num):
            primes.append(num)
    
    return primes


# Test cases
if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Find nth Prime Number")
    print("=" * 60)
    
    # Test Case 1 - 1st prime
    n1 = 1
    result1 = nth_prime_brute_force(n1)
    print(f"n = {n1}")
    print(f"{n1}st prime number: {result1}")
    
    # Test Case 2 - 5th prime
    n2 = 5
    result2 = nth_prime_brute_force(n2)
    result2_sieve = nth_prime_sieve_optimized(n2)
    print(f"\nn = {n2}")
    print(f"{n2}th prime number (brute force): {result2}")
    print(f"{n2}th prime number (sieve): {result2_sieve}")
    
    # Test Case 3 - 10th prime
    n3 = 10
    result3 = nth_prime_brute_force(n3)
    result3_sieve = nth_prime_sieve_optimized(n3)
    print(f"\nn = {n3}")
    print(f"{n3}th prime number (brute force): {result3}")
    print(f"{n3}th prime number (sieve): {result3_sieve}")
    
    # Test Case 4 - First 10 primes
    n4 = 10
    primes_list = get_first_n_primes(n4)
    print(f"\nFirst {n4} prime numbers:")
    print(primes_list)
    
    # Test Case 5 - 20th prime
    n5 = 20
    result5 = nth_prime_brute_force(n5)
    print(f"\nn = {n5}")
    print(f"{n5}th prime number: {result5}")
    
    # Test Case 6 - Primes in range
    start = 10
    end = 50
    primes_range = get_primes_in_range(start, end)
    print(f"\nPrime numbers in range [{start}, {end}]:")
    print(primes_range)
    print(f"Count: {len(primes_range)}")
    
    # Test Case 7 - 100th prime
    n7 = 100
    result7 = nth_prime_sieve_optimized(n7)
    print(f"\nn = {n7}")
    print(f"{n7}th prime number: {result7}")

File: basic_dsa_thirty_three.java
/**
 * Basic DSA Problems - One More Fundamental Question
 * Simple and essential DSA problems for beginners
 */
import java.util.*;

public class basic_dsa_thirty_three {
    
    // Problem: Find nth Prime Number
    // Find the nth prime number in the sequence of prime numbers
    // Example: 1st prime = 2, 2nd prime = 3, 3rd prime = 5, 4th prime = 7, 5th prime = 11
    
    /**
     * Check if a number is prime
     * Time Complexity: O(√n)
     * Space Complexity: O(1)
     */
    public static boolean isPrime(int n) {
        if (n < 2) {
            return false;
        }
        if (n == 2) {
            return true;
        }
        if (n % 2 == 0) {
            return false;
        }
        
        for (int i = 3; i * i <= n; i += 2) {
            if (n % i == 0) {
                return false;
            }
        }
        
        return true;
    }
    
    /**
     * Find nth prime number using brute force
     * Time Complexity: O(n * √n)
     * Space Complexity: O(1)
     */
    public static int nthPrimeBruteForce(int n) {
        if (n <= 0) {
            throw new IllegalArgumentException("n must be positive");
        }
        
        if (n == 1) {
            return 2;
        }
        
        int count = 1;
        int num = 3;
        
        while (count < n) {
            if (isPrime(num)) {
                count++;
                if (count == n) {
                    return num;
                }
            }
            num += 2; // Check only odd numbers (except 2)
        }
        
        return num - 2;
    }
    
    /**
     * Find nth prime number using optimized approach with Sieve of Eratosthenes
     * Time Complexity: O(n log log n) for sieve
     * Space Complexity: O(n)
     */
    public static int nthPrimeSieveOptimized(int n) {
        if (n <= 0) {
            throw new IllegalArgumentException("n must be positive");
        }
        
        if (n == 1) {
            return 2;
        }
        
        // Estimate upper bound: n * log(n) + n * log(log(n)) for n > 6
        if (n <= 6) {
            int[] primes = {2, 3, 5, 7, 11, 13};
            return primes[n - 1];
        }
        
        int upperBound = (int) (n * Math.log(n) + n * Math.log(Math.log(n)));
        if (upperBound < 20) {
            upperBound = 20;
        }
        
        // Sieve of Eratosthenes
        boolean[] isPrimeList = new boolean[upperBound + 1];
        Arrays.fill(isPrimeList, true);
        isPrimeList[0] = isPrimeList[1] = false;
        
        for (int i = 2; i * i <= upperBound; i++) {
            if (isPrimeList[i]) {
                for (int j = i * i; j <= upperBound; j += i) {
                    isPrimeList[j] = false;
                }
            }
        }
        
        // Collect primes
        int count = 0;
        for (int i = 2; i <= upperBound; i++) {
            if (isPrimeList[i]) {
                count++;
                if (count == n) {
                    return i;
                }
            }
        }
        
        // If not enough primes found, use brute force
        return nthPrimeBruteForce(n);
    }
    
    /**
     * Get list of first n prime numbers
     * Time Complexity: O(n * √n)
     * Space Complexity: O(n)
     */
    public static ArrayList<Integer> getFirstNPrimes(int n) {
        ArrayList<Integer> primes = new ArrayList<>();
        if (n <= 0) {
            return primes;
        }
        
        int num = 2;
        
        while (primes.size() < n) {
            if (isPrime(num)) {
                primes.add(num);
            }
            num++;
        }
        
        return primes;
    }
    
    /**
     * Get all prime numbers in a given range
     * Time Complexity: O((end - start) * √n)
     * Space Complexity: O(k) where k is the number of primes
     */
    public static ArrayList<Integer> getPrimesInRange(int start, int end) {
        ArrayList<Integer> primes = new ArrayList<>();
        
        for (int num = start; num <= end; num++) {
            if (isPrime(num)) {
                primes.add(num);
            }
        }
        
        return primes;
    }
    
    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Find nth Prime Number");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - 1st prime
        int n1 = 1;
        int result1 = nthPrimeBruteForce(n1);
        System.out.println("n = " + n1);
        System.out.println(n1 + "st prime number: " + result1);
        
        // Test Case 2 - 5th prime
        int n2 = 5;
        int result2 = nthPrimeBruteForce(n2);
        int result2_sieve = nthPrimeSieveOptimized(n2);
        System.out.println("\nn = " + n2);
        System.out.println(n2 + "th prime number (brute force): " + result2);
        System.out.println(n2 + "th prime number (sieve): " + result2_sieve);
        
        // Test Case 3 - 10th prime
        int n3 = 10;
        int result3 = nthPrimeBruteForce(n3);
        int result3_sieve = nthPrimeSieveOptimized(n3);
        System.out.println("\nn = " + n3);
        System.out.println(n3 + "th prime number (brute force): " + result3);
        System.out.println(n3 + "th prime number (sieve): " + result3_sieve);
        
        // Test Case 4 - First 10 primes
        int n4 = 10;
        ArrayList<Integer> primesList = getFirstNPrimes(n4);
        System.out.println("\nFirst " + n4 + " prime numbers:");
        System.out.println(primesList);
        
        // Test Case 5 - 20th prime
        int n5 = 20;
        int result5 = nthPrimeBruteForce(n5);
        System.out.println("\nn = " + n5);
        System.out.println(n5 + "th prime number: " + result5);
        
        // Test Case 6 - Primes in range
        int start = 10;
        int end = 50;
        ArrayList<Integer> primesRange = getPrimesInRange(start, end);
        System.out.println("\nPrime numbers in range [" + start + ", " + end + "]:");
        System.out.println(primesRange);
        System.out.println("Count: " + primesRange.size());
        
        // Test Case 7 - 100th prime
        int n7 = 100;
        int result7 = nthPrimeSieveOptimized(n7);
        System.out.println("\nn = " + n7);
        System.out.println(n7 + "th prime number: " + result7);
    }
}
